observatory-dags/observatory/dags/workflows/onix_workflow.py

```diff
@@ -0,0 +1,509 @@
+    """
+
+    DAG_ID_PREFIX = "onix_workflow"
+
+    def __init__(
+        self,
+        *,
+        org_name: str,
+        gcp_project_id: str,
+        gcp_bucket_name: str,
+        telescope_sensor: ExternalTaskSensor,
+        onix_dataset_id: str = "onix",
+        onix_table_id: str = "onix",
+        dag_id: Optional[str] = None,
+        start_date: Optional[datetime] = datetime(2021, 3, 28),
+        schedule_interval: Optional[str] = "@weekly",
+        catchup: Optional[bool] = False,
+        data_partners: List[OaebuPartners] = list(),
+    ):
+        """Initialises the workflow object.
+        :param org_name: Organisation name.
+        :param gcp_project_id: Project ID in GCP.
+        :param gcp_bucket_name: GCP bucket name to store files.
+        :param telescope_sensor: Telescope sensor to monitor.
+        :param onix_dataset_id: GCP dataset ID of the onix data.
+        :param onix_table_id: GCP table ID of the onix data.
+        :param dag_id: DAG ID.
+        :param start_date: Start date of the DAG.
+        :param schedule_interval: Scheduled interval for running the DAG.
+        :param catchup: Whether to catch up missed DAG runs.
+        :param data_partners: OAEBU data sources.
+        """
+
+        self.dag_id = dag_id
+        if dag_id is None:
+            self.dag_id = make_dag_id(self.DAG_ID_PREFIX, org_name)
+
+        self.org_name = org_name
+        self.gcp_project_id = gcp_project_id
+        self.gcp_bucket_name = gcp_bucket_name
+        self.onix_dataset_id = onix_dataset_id
+        self.onix_table_id = onix_table_id
+
+        # Initialise Telesecope base class
+        super().__init__(
+            dag_id=self.dag_id, start_date=start_date, schedule_interval=schedule_interval, catchup=catchup
+        )
+
+        # Wait on ONIX telescope to finish
+        self.add_sensor(telescope_sensor)
+
+        # Aggregate Works
+        self.add_task(self.aggregate_works)
+        self.add_task(self.upload_aggregation_tables)
+        self.add_task(self.bq_load_workid_lookup)
+        self.add_task(self.bq_load_workid_lookup_errors)
+        self.add_task(self.bq_load_workfamilyid_lookup)
+
+        # Create OAEBU Intermediate tables
+        self.create_oaebu_intermediate_table_tasks(data_partners)
+
+        # Cleanup tasks
+        self.add_task(self.cleanup)
+
+    def make_release(self, **kwargs) -> OnixWorkflowRelease:
+        """Creates a release object.
+        :param kwargs: From Airflow. Contains the execution_date.
+        :return: OnixWorkflowRelease object.
+        """
+
+        release_date = kwargs["execution_date"]
+        release = OnixWorkflowRelease(
+            dag_id=self.dag_id,
+            release_date=release_date,
+            gcp_project_id=self.gcp_project_id,
+            gcp_bucket_name=self.gcp_bucket_name,
+            onix_dataset_id=self.onix_dataset_id,
+            onix_table_id=self.onix_table_id,
+        )
```

Comment:
The biggest thing that will need changing is refactoring the OnixWorkflow to process multiple releases. The OnixTelescope processes multiple releases for each DAG run (e.g. if multiple ONIX files have been deposited), therefore the OnixWorkflow should also be able to process multiple releases.

I'm thinking that a better way to communicate what releases need to be processed by the OnixWorkflow, is to pull the XCom published by the OnixTelescope for the same period. This XCom contains a list of release dates, e.g. `[{'release_date': <Pendulum [2021-04-20T00:00:00+00:00]>, 'file_name': '20210420_ANUPRESS_ONIX.xml'}]`. This means that you would not use `select_table_suffixes` further down, just the release dates pulled from XCom stored in the release object.

You can pull them like so:
```python
ti: TaskInstance = kwargs['ti']
records = ti.xcom_pull(
    dag_id='onix_telescope_name',
    key=OnixTelescope.RELEASE_INFO,
    task_ids='list_release_info',
    include_prior_dates=False)
```